### PR TITLE
Add InterpolateBackground to SNSPowderReduction

### PR DESCRIPTION
**Pull request by ghost**
**Do not merge before [PR 36010](https://github.com/mantidproject/mantid/pull/36010)**

**Description of work**

This PR adds changes to the `Powder Diffraction Reduction` GUI interface to include the parameters for Interpolating the temperature of background runs. This PR also updates `SNSPowderReduction` to include the new parameters and perform the background interpolation. 

These changes are for [EWM 892](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20%28Change%20Management%29#action=com.ibm.team.workitem.viewWorkItem&id=892).

** Updated GUI **
image

**To test:**
All Existing tests should pass, unit tests were previously included to independently test correct functionality of `InterpolateBackground`

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py [hunk 1]
```diff
@@ -337,6 +337,11 @@ def PyInit(self):
         )
 
         self.copyProperties("AlignAndFocusPowderFromFiles", ["FrequencyLogNames", "WaveLengthLogNames"])
+        self.declareProperty(
+            "InterpolateTargetTemp",
+            0.0,
```
Comment (rboston628) [inline Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py:342]: What units are used for temperature?

File: Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py [hunk 2]
```diff
@@ -418,6 +425,11 @@ def PyExec(self):  # noqa
         self._elementSize = self.getProperty("ElementSize").value
         self._num_wl_bins = self.getProperty("NumWavelengthBins").value
 
+        self._interpoTemp = self.getProperty("InterpolateTargetTemp").value
+        self._enableInterpo = False
+        if self._interpoTemp > 0.0:
+            self._enableInterpo = True
+
```
Comment (rboston628) [inline Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py:432]:
``` python
self._enableInterpo = (self._interpoTemp > 0.0)
```
Are we sure interpoTemp cannot be normally non-positive?

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/reduction_gui/widgets/diffraction/diffraction_run_setup.py
Comment (rboston628) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/reduction_gui/widgets/diffraction/diffraction_run_setup.py:1]: Are we sure this story requires changing the GUI?

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/ui/diffraction/diffraction_run_setup.ui
Comment (rboston628) [inline qt/python/mantidqtinterfaces/mantidqtinterfaces/ui/diffraction/diffraction_run_setup.ui:1]: These changes seem to be largely reorganizing this xml, or changing row numbers.  Are these changes necessary?

File: Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py [hunk 3]
```diff
@@ -1398,7 +1413,28 @@ def _process_container_runs(self, can_run_numbers, samRunIndex, preserveEvents,
...
+                    InputWorkspace=interpo_ws,
+                    OutputWorkspace=interpo_ws,
+                    Filter="Butterworth",
+                    Params=smoothParams,
+                    IgnoreXBins=True,
+                    AllSpectra=True,
+                )
+            return interpo_ws
+        else:
+            can_run_ws_name, can_run_number = self._generate_container_run_name(can_run_numbers, samRunIndex)
```
Comment (rboston628) [inline Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py:1437]: Are there any tests of correct performance of this new behavior?

File: Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py (hunk 1 above)
Comment (searscr) [reply Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py:342]: Kelvin, I will add that to the doc

File: Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py (hunk 2 above)
Comment (searscr) [reply Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py:432]: Temp is in Kelvin, 0 is an appropriate floor

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/reduction_gui/widgets/diffraction/diffraction_run_setup.py
Comment (searscr) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/reduction_gui/widgets/diffraction/diffraction_run_setup.py:1]: Story specifically asks for GUI changes. GUI has been changed to match requested changes.

File: qt/python/mantidqtinterfaces/mantidqtinterfaces/ui/diffraction/diffraction_run_setup.ui
Comment (searscr) [reply qt/python/mantidqtinterfaces/mantidqtinterfaces/ui/diffraction/diffraction_run_setup.ui:1]: Changes reflect how the items are organized in the QT widget

File: Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py (hunk 3 above)
Comment (searscr) [reply Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py:1437]: This is one small portion of a larger algorithm for reduction. I was having a hard time coming up with test to specifically check this functionality

Comment (rboston628) [reply Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py:1437]: The new functionality here is only an `if` clause.  The behavior inside the `if` was completed inside PR #35977 , and has associated tests.

**Comment by Kvieta1990** [comment]
I grabbed the PR and did a local test. The result looks very positive! Here follows are details.

I used the two XML files included in the attached zip file for the testing -- one for using the exact temperature measurement at 300 K as the empty container and one for using the interpolated empty container (200 & 400 K measurements interpolated to 300 K). The result can be found in the two xye files which shows very consistent outcome, as shown in the figure below.

[test.zip](https://github.com/mantidproject/mantid/files/12503259/test.zip)

POWGEN_Bkg_Interpo_Test

**Comment by rboston628** [comment]
The update meets standards for python code, and extends the functionality to improve the use with instruments.  The GUI changes have been discussed and designed to be most clear for users.  Local testing by instrument scientists shows the changes to work as desired.

**Review by rboston628 — DISMISSED** [review]

**Comment by Kvieta1990** [comment]
When testing out the latest implementation, it seems that the GUI does not launch, throwing out the following error,

```
Traceback (most recent call last):
  File "/SNS/snfs1/instruments/NOM/shared/Dev/mantid/qt/applications/workbench/workbench/app/mainwindow.py", line 403, in <lambda>
    action.triggered.connect(lambda checked_py, script=script: self.launch_custom_python_gui(script))
  File "/SNS/snfs1/instruments/NOM/shared/Dev/mantid/qt/applications/workbench/workbench/app/mainwindow.py", line 362, in launch_custom_python_gui
    self.interface_executor.execute(open(filename).read(), filename)
  File "/SNS/snfs1/instruments/NOM/shared/Dev/mantid/qt/python/mantidqt/mantidqt/widgets/codeeditor/execution.py", line 153, in execute
    executor.execute(code_str, filename, flags, self.globals_ns, line_offset)
  File "/SNS/snfs1/instruments/NOM/shared/Dev/mantid/qt/python/mantidqtinterfaces/mantidqtinterfaces/Powder_Diffraction_Reduction.py", line 32, in <module>
    reducer.open_file(autopath)
  File "/SNS/snfs1/instruments/NOM/shared/Dev/mantid/qt/python/mantidqtinterfaces/mantidqtinterfaces/reduction_application.py", line 418, in open_file
    self._interface.load_file(file_path)
  File "/SNS/snfs1/instruments/NOM/shared/Dev/mantid/qt/python/mantidqtinterfaces/mantidqtinterfaces/reduction_gui/instruments/interface.py", line 102, in load_file
    self.scripter.from_xml(file_name)
  File "/SNS/NOM/shared/Dev/mantid/scripts/reduction_gui/reduction/scripter.py", line 539, in from_xml
    item.state().from_xml(xml_str)
  File "/SNS/NOM/shared/Dev/mantid/scripts/reduction_gui/reduction/diffraction/diffraction_run_setup_script.py", line 249, in from_xml
    self.interpolatetemp = BaseScriptElement.getFloatElement(
  File "/SNS/NOM/shared/Dev/mantid/scripts/reduction_gui/reduction/scripter.py", line 157, in getFloatElement
    return float(value)
ValueError: could not convert string to float: ''
```

Here are two XML files that both cause this error,

[XML_files_giving_error.zip](https://github.com/mantidproject/mantid/files/12526478/XML_files_giving_error.zip)

The one with `old` in the filename is the saved XML from the current production version of the GUI (Mantidworkbench 6.7.0). The other one is the saved XML from the GUI associated with current PR. Since initially, the GUI would crash with the existing XML file in my local configuration area, I had to move the local version to somewhere else and then the GUI could launch properly. Then I populated contents as usual and saved the config to XML file and that is what is attached here. But as I said, even for the version of XML file saved from itself, the new version of GUI still could not launch the XML.

**Comment by Kvieta1990** [comment]
Unfortunately the latest codes will crash the reduction, with the configuration file as below,

[snspowderreduction_interpo.zip](https://github.com/mantidproject/mantid/files/12531914/snspowderreduction_interpo.zip)

Here is the error,

```
Minus: Plus::performEventBinaryOperation() cannot subtract a histogram from an event list in an EventWorkspace. Try switching to a Workspace2D before using Minus.
Error in execution of algorithm Minus:
Minus: error (see log)
Error in execution of algorithm SNSPowderReduction:
Minus-v1: Minus: error (see log)
  at line 637 in '/SNS/NOM/shared/Dev/mantid/Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py'
  caused by line 1076 in '/SNS/snfs1/instruments/NOM/shared/Dev/mantid/Framework/PythonInterface/mantid/simpleapi.py'
Failed to execute script: RuntimeError on line 1076 of '/SNS/snfs1/instruments/NOM/shared/Dev/mantid/Framework/PythonInterface/mantid/simpleapi.py': SNSPowderReduction-v1: Minus-v1: Minus: error (see log)
  at line 637 in '/SNS/NOM/shared/Dev/mantid/Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py'
  caused by line 1076 in '/SNS/snfs1/instruments/NOM/shared/Dev/mantid/Framework/PythonInterface/mantid/simpleapi.py'
Traceback (most recent call last):
  File "/SNS/NOM/shared/Dev/mantid/scripts/reduction_gui/reduction/scripter.py", line 613, in on_error
    raise RuntimeError(str(arg))
RuntimeError: RuntimeError on line 1076 of '/SNS/snfs1/instruments/NOM/shared/Dev/mantid/Framework/PythonInterface/mantid/simpleapi.py': SNSPowderReduction-v1: Minus-v1: Minus: error (see log)
  at line 637 in '/SNS/NOM/shared/Dev/mantid/Framework/PythonInterface/plugins/algorithms/SNSPowderReduction.py'
  caused by line 1076 in '/SNS/snfs1/instruments/NOM/shared/Dev/mantid/Framework/PythonInterface/mantid/simpleapi.py'
```

**Comment by Kvieta1990** [comment]
With the latest commit, everything is working as expected and I think we are ready to go. Thank you @searscr!

**Review by Kvieta1990 — COMMENTED** [review]
I will approve the PR once the comment from @sf1919 is addressed.

**Comment by sf1919** [comment]
Unfortunately the build package from branch run I asked for failed. However we think this was a problem with the node so we've taken it offline to investigate and kicked off a new run for this PR - https://builds.mantidproject.org/job/build_packages_from_branch/567/

**Comment by sf1919** [comment]
This build has now passed so there is nothing blocking approval. Thanks!

**Review by Kvieta1990 — APPROVED** [review]
